### PR TITLE
Add configuration for DFL in JSON format

### PR DESCRIPTION
**Pull request by lecourtoisflo**

### CODE REVIEW

File: sources/Context.cpp [hunk 1]
```diff
@@ -17,11 +17,26 @@
...
+Context::Context(const std::string& networkFilepath, const std::string& configFilepath) :
+    networkManager_(networkFilepath),
+    config_(configFilepath),
+    slack_node_{} {
+  auto automatic_slack_node = networkManager_.getSlackNode();
+  if (automatic_slack_node.is_initialized() && !config_.isAutomaticSlackBusOn()) {
+    slack_node_ = *automatic_slack_node;
+  } else {
+    // automatic slack node not given: we must compute it ourselves
+    if (!automatic_slack_node.is_initialized() && config_.isAutomaticSlackBusOn()) {
```
Comment (cossartque) [inline sources/Context.cpp:34]: Maybe change the name "automatic_slack_node" to "found_slack_node" ?

File: sources/Context.cpp [hunk 2]
```diff
@@ -17,11 +17,26 @@
...
+      // TODO(lecourtoisflo) use warning without restriction for language
+      LOG(warn) << "Automatic slack node requested but not present in network input file" << LOG_ENDL;
```
Comment (cossartque) [inline sources/Context.cpp:36]: You should warn the user when Automatic=false and Found=false (in that case you warn and you calculate the slack node automatically)
If Automatic=true then calculate (whatever if Found or not)
If Automatic=false and Found=true, pick the slack found in the iidm

File: sources/Context.cpp [hunk 3]
```diff
@@ -17,11 +17,26 @@
...
 namespace dfl {
-Context::Context(const std::string& networkFilepath) : networkManager_(networkFilepath) {
-  networkManager_.onNode(std::bind(&Context::processSlackNode, this, std::placeholders::_1));
+Context::Context(const std::string& networkFilepath, const std::string& configFilepath) :
+    networkManager_(networkFilepath),
+    config_(configFilepath),
+    slack_node_{} {
+  auto automatic_slack_node = networkManager_.getSlackNode();
+  if (automatic_slack_node.is_initialized() && !config_.isAutomaticSlackBusOn()) {
+    slack_node_ = *automatic_slack_node;
```
Comment (cossartque) [inline sources/Context.cpp:31]: found_slack_node ?

**Review by cossartque — CHANGES_REQUESTED** [review]

File: sources/Context.cpp (hunk 1 above)
Comment (lecourtoisflo) [reply sources/Context.cpp:34]: Done

File: sources/Context.cpp (hunk 2 above)
Comment (lecourtoisflo) [reply sources/Context.cpp:36]: OK

File: sources/Context.cpp (hunk 3 above)
Comment (lecourtoisflo) [reply sources/Context.cpp:31]: OK

File: sources/Context.cpp [hunk 4]
```diff
@@ -17,11 +17,26 @@
...
+    config_(configFilepath),
+    slack_node_{} {
+  auto found_slack_node = networkManager_.getSlackNode();
+  if (found_slack_node.is_initialized() && !config_.isAutomaticSlackBusOn()) {
+    slack_node_ = *found_slack_node;
+  } else {
+    // slack node not given in iidm or not requested: we must compute it ourselves
+    if (!found_slack_node.is_initialized() && !config_.isAutomaticSlackBusOn()) {
+      // TODO(lecourtoisflo) use warning without restriction for language
+      LOG(warn) << "Automatic slack node requested but not present in network input file" << LOG_ENDL;
```
Comment (cossartque) [inline sources/Context.cpp:36]: I would say "Slack node requested but no present in network input file. It is computed automatically"

Comment (cossartque) [inline sources/Context.cpp:36]: In addition to the LOG(warn), don't you need to choose the slack node ? I would expect something like slack_node_ = *computed_slack_node

File: sources/Context.cpp [hunk 5]
```diff
@@ -17,11 +17,26 @@
...
+    }
```
Comment (cossartque) [inline sources/Context.cpp:37]: The case if (found_slack_node.is_initialized() && config_.isAutomaticSlackBusOn()) is missing I think

**Review by cossartque — CHANGES_REQUESTED** [review]

File: sources/Context.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply sources/Context.cpp:36]: done next line

File: sources/Context.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply sources/Context.cpp:37]: Done next line

File: sources/Context.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply sources/Context.cpp]: OK

File: sources/Inputs/include/Configuration.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+   * @brief determines ifLCC are considered as loads
+   *
+   * @returns the parameter value
+   */
+  bool useLCCAsLoads() const {
+    return LCCAsLoads_;
+  }
+
+ private:
+  bool InfiniteReactiveLimits_ = false;  ///< infinite reactive limits
```
Comment (gautierbureau) [inline sources/Inputs/include/Configuration.h:101]: attribute start with a lowercase

File: tests/inputs/res/config_default.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "dfl-config": {}
+}
```
Comment (gautierbureau) [inline tests/inputs/res/config_default.json:3]: newline

File: tests/inputs/res/config.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  "dfl-config": {
+    "InfiniteReactiveLimits": "true",
+    "PSTRegulationOn": "false",
+    "SVCRegulationOn": "false",
+    "ShuntRegulationOn": "false",
+    "AutomaticSlackBusOn": "false",
+    "VSCAsGenerators": "true",
+    "LCCAsLoads": "true"
+  }
+}
```
Comment (gautierbureau) [inline tests/inputs/res/config.json:11]: newline

File: tests/main/res/config.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  "dfl-config": {
+    "InfiniteReactiveLimits": "true",
+    "PSTRegulationOn": "false",
+    "SVCRegulationOn": "false",
+    "ShuntRegulationOn": "false",
+    "AutomaticSlackBusOn": "false",
+    "VSCAsGenerators": "true",
+    "LCCAsLoads": "true"
+  }
+}
```
Comment (gautierbureau) [inline tests/main/res/config.json:11]: newline

File: tests/main/CMakeLists.txt [hunk 10]
```diff
@@ -10,5 +10,7 @@
 add_test(
   # Use some per-module/project prefix so that it is easier to run only tests for this module
   NAME MAIN.launch
-  COMMAND DynaflowLauncher
+  COMMAND DynaflowLauncher --iidm=res/IEEE14.iidm --config=res/config.json
```
Comment (gautierbureau) [inline tests/main/CMakeLists.txt:13]: I would have put an uppercase to Flow

**Review by gautierbureau — APPROVED** [review]

File: sources/Inputs/include/Configuration.h (hunk 6 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Configuration.h:101]: OK

File: tests/inputs/res/config_default.json (hunk 7 above)
Comment (lecourtoisflo) [reply tests/inputs/res/config_default.json:3]: not relevant for JSON files

File: tests/inputs/res/config.json (hunk 8 above)
Comment (lecourtoisflo) [reply tests/inputs/res/config.json:11]: not relevant for JSON files

File: tests/main/res/config.json (hunk 9 above)
Comment (lecourtoisflo) [reply tests/main/res/config.json:11]: not relevant for JSON files

**Review by cossartque — APPROVED** [review]

File: sources/Context.cpp [hunk 11]
```diff
@@ -17,11 +17,26 @@
...
+  auto found_slack_node = networkManager_.getSlackNode();
+  if (found_slack_node.is_initialized() && !config_.isAutomaticSlackBusOn()) {
+    slack_node_ = *found_slack_node;
+  } else {
+    // slack node not given in iidm or not requested: it is computed internally
+    if (!found_slack_node.is_initialized() && !config_.isAutomaticSlackBusOn()) {
+      // TODO(lecourtoisflo) use warning without restriction for language
+      LOG(warn) << "Network slack node requested but not present in network input file" << LOG_ENDL;
+    }
+    networkManager_.onNode(std::bind(&Context::processSlackNode, this, std::placeholders::_1));
```
Comment (rosiereflo) [inline sources/Context.cpp:38]: Would be good to display the final slack node and potentially its origin (automatic, from iidm ...)

Comment (rosiereflo) [reply sources/Context.cpp:38]: (we can start with an english message and add later dictionaries)

File: sources/Inputs/src/NetworkManager.cpp [hunk 12]
```diff
@@ -100,5 +100,11 @@ NetworkManager::walkNodes() {
   }
 }
 
+boost::optional<boost::shared_ptr<Node>>
+NetworkManager::getSlackNode() const {
+  // for now, we cannot retrieve from dynawo data interface this information: we will in a future version
```
Comment (rosiereflo) [inline sources/Inputs/src/NetworkManager.cpp:105]: Please add a todo so that we can retrieve easily open subjects

**Review by rosiereflo — CHANGES_REQUESTED** [review]

**Review by marcochiaramello — APPROVED** [review]

File: sources/Context.cpp (hunk 11 above)
Comment (lecourtoisflo) [reply sources/Context.cpp:38]: Will be done in future patch (when dictionnary is added)

File: sources/Inputs/src/NetworkManager.cpp (hunk 12 above)
Comment (lecourtoisflo) [reply sources/Inputs/src/NetworkManager.cpp:105]: OK

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**